core/src/main/scala/chisel3/Aggregate.scala

```diff
@@ -591,6 +591,67 @@ object VecInit extends SourceInfoDoc {
   def do_tabulate[T <: Data](n: Int)(gen: (Int) => T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Vec[T] =
     apply((0 until n).map(i => gen(i)))
 
+  /** Creates a new 2D [[Vec]] of length `n by m` composed of the results of the given
+    * function applied over a range of integer values starting from 0.
+    *
+    * @param n number of 1D vectors inside outer vector
+    * @param m number of elements in each 1D vector (the function is applied from
+    * 0 to `n-1`)
+    * @param gen function that takes in an Int (the index) and returns a
+    * [[Data]] that becomes the output element
+    */
+  def tabulate[T <: Data](n: Int, m: Int)(gen: (Int) => T): Vec[Vec[T]] = macro VecTransform.tabulate2D
+
+  /** @group SourceInfoTransformMacro */
+  def do_tabulate[T <: Data](n: Int, m: Int)(gen: (Int) => T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Vec[Vec[T]] = {
+    val myVec = Wire(Vec(n, Vec(m, cloneSupertype((0 until m).map(i => gen(i)), "Vec"))))
+    myVec.foreach {
+      _ := VecInit.tabulate(m)(gen)
+    }
+    myVec
+  }
```

Comment:
Using `:=` is actually not correct--`VecInit.tabulate` needs to support bidirectional aggregates and `:=` is a unidirectional connection. Now unfortunately, just switching this implementation to using `<>` also won't work due to the fact that `<>` is commutative and thus it is ambiguous which direction connections should go if both sides of a `<>` are wires (if one side is a port, that side "anchors" the connection making it unambiguous). For example:

[This](https://scastie.scala-lang.org/E8fkK8CKTzKrEBSI4ZprdQ) works (both sides ports)
```scala
  val enq = IO(Vec(4, Decoupled(UInt(8.W))))
  val deq = IO(Flipped(Vec(4, Decoupled(UInt(8.W)))))
  deq <> enq
```

[This](https://scastie.scala-lang.org/YuQsQaVTQ1aHzkxpqEhUvA) works (one side port, one side wire)
```scala
  val enq = IO(Vec(4, Decoupled(UInt(8.W))))
  val deq = IO(Flipped(Vec(4, Decoupled(UInt(8.W)))))
  // Effectively what VecInit does
  val wire = Wire(Vec(4, Decoupled(UInt(8.W))))
  wire <> enq
  
  deq <> wire
```

But [this](https://scastie.scala-lang.org/rEH7I0kIR8WmVH5rZK1yPw) doesn't work (both sides wire):
```scala
  val enq = IO(Vec(4, Decoupled(UInt(8.W))))
  val deq = IO(Flipped(Vec(4, Decoupled(UInt(8.W)))))
  // Effectively what VecInit does
  val wire = Wire(Vec(4, Decoupled(UInt(8.W))))
  wire <> enq
  val wire2 = Wire(Vec(4, Flipped(Decoupled(UInt(8.W)))))
  wire2 <> deq
  
  wire2 <> wire
```

In order to make this higher dimensional `fills` and `tabulates` work, they cannot create a Wire and delegate to a function that also creates a Wire. I think the only way to do this correctly is to create the wire in each `tabulate` and then connect every single element (no delegation).

Also, looking at the fact that the 1D tabulate delegates to `do_apply` which has [this weird logic](https://github.com/chipsalliance/chisel3/blob/713de6b823d8707246935b9e31ed2fbafeaeca32/core/src/main/scala/chisel3/Aggregate.scala#L552), I think we should preserve that functionality as well. I'm envisioning factoring that out into a private higher-order function like so:

```scala
private def getConnectOpFromDirectionality[T <: Data](proto: T): (T, T) => Unit = proto.direction match {
  case ActualDirection.Input | ActualDirection.Output | ActualDirection.Unspecified =>
    // When internal wires are involved, driver / sink must be specified explicitly, otherwise
    // the system is unable to infer which is driver / sink
    (x, y) => x := y
  case ActualDirection.Bidirectional(_) =>
    // For bidirectional, must issue a bulk connect so subelements are resolved correctly.
    // Bulk connecting two wires may not succeed because Chisel frontend does not infer
    // directions.
    (x, y) => x <> y
}
```

Then you could use that for this function as:
```suggestion
  // Also gen should take more Ints, 1 per dimensionality
  def do_tabulate[T <: Data](n: Int, m: Int)(gen: (Int, Int) => T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Vec[Vec[T]] = {
    // TODO make this lazy (requires LazyList and cross compilation, beyond the scope of this PR)
    val elts = Seq.tabulate(n, m)(gen)
   
    val flatElts = elts.flatten
    
    // These checks are copied from do_apply, perhaps some room for factoring out
    require(flatElts.nonEmpty, "Vec hardware values are not allowed to be empty")
    flatElts.foreach(requireIsHardware(_, "vec element"))
    
    val tpe = cloneSupertype(flatElts, "Vec.tabulate")
    val myVec = Wire(Vec(n, Vec(m, tpe)))
    
    val op = getConnectOpFromDirectionality(tpe)
    // We could get fancy here to have reuse across the multiple dimensions, but being more verbose is more clear
    for {
      (xs1D, ys1D) <- myVec zip elts
      (x, y) <- xs1D zip ys1D
    } do op(x, y)
    myVec
  }
```

I know this is a lot, and this is made complex by the lack of a non-commutative `<>`, but it is what it is.

Reply:
Instead of the nested `for`, we could actually just flatten and iterate:
```scala
for ((x, y) <- myVec.flatten zip flatElts) {
  op(x, y)
}
```
This requires fewer intermediate names and would work basically the same way for the higher dimensional versions (with extra `.flatten`s)